proposals/p3848.md

```diff
@@ -0,0 +1,587 @@
+syntactic challenge in the form of cliffs, explained below.
+
+Terse lambdas were slated to be the most compact form of a lambda. Combined with
+a [sigil introducer](#introducer), they would be syntactically minimal. One way
+in which syntax was minimized was the granting of an **implicit** default
+[capture](#function-captures) mode. If no square brackets were present, by-value
+captures would be allowed. This, combined with the lack of an arrow to signify a
+return value, created syntax of the following form (being passed into the filter
+function below).
+
+```
+let zero: i32 = 0;
+let list_all: List(i32) = GetAllValues();
+let list_positive: List(i32) = list_all.Filter(
+  @ $0 > zero
+);
+```
+
+To give users more control over the feature set in a lambda, the next step up
+was an elaborated lambda. This provided the ability to add both square brackets
+and explicit parameters to lambdas at the cost of more syntax. Unfortunately,
+this also meant there was a bit of a syntactic cliff and a stumbling block. It
+was considered desirable for empty square brackets to mean capturing is
+disabled. But since the no-square-brackets form needed to support capturing for
+terse lambdas, elaborated lambdas needed to both add the square brackets and
+also add an explicit default capture mode at the same time just to maintain the
+existing capturing behavior. The net result was code that looked like the
+following (being passed into the filter function again).
+
+```
+let zero: i32 = 0;
+let list_all: List(i32) = GetAllValues();
+let list_positive: List(i32) = list_all.Filter(
+  @[let](x: auto) x > zero
+);
+```
+
+Finally, if a user wanted to upgrade a lambda to a function declaration, this
+created another cliff where they needed to switch from the sigil to the `fn`
+keyword, on top of adding a name. Ultimately these downsides suggested that a
+continuous syntax was the better path forward, even though the minimal lambda
+syntax would be a bit less terse.
+
+## Introducer
+
+**Proposal**: Introduce with the `fn` keyword to mirror function declarations.
+If a statement or declaration begins with `fn`, a name is required and it
+becomes a function declaration. Otherwise, if in an expression context, `fn`
+introduces a lambda.
+
+```
+let lambda1: auto = fn => T.Make();
+
+let lambda2: auto = fn []() -> T { return T.Make(); };
+
+fn FunctionDeclaration1 => T.Make();
+
+fn FunctionDeclaration2[]() -> T { return T.Make(); }
+```
+
+### Alternative Considered: Sigil
+
+Introduce with a sigil, such as `$` or `@`. Since introducer punctuation is such
+a scarce resource, and since there was no consensus on what sigil would best
+represent a lambda, and since there was a desire to create a more continuous
+syntax between lambdas and function declarations, this alternative was decided
+against.
+
+```
+let lambda1: auto = @ => T.Make();
+
+let lambda2: auto = @[]() -> T { return T.Make(); };
+```
+
+## Positional Parameters
```

Comment:
As I think Swift is one of the few other languages that has a similar feature, it would be good to a) mention that and b) indicate what if any things really differ between them.